milli/src/search/query_tree.rs

```diff
@@ -562,16 +590,19 @@ fn create_matching_words(
                     for synonym in synonyms {
                         let synonym = synonym
                             .into_iter()
-                            .map(|syn| MatchingWord::new(syn, 0, false))
+                            .flat_map(|syn| matching_word_cache.insert(syn, 0, false))
                             .collect();
```

Comment:
Now, `matching_word_cache.insert` can return `None` if the length of its string argument exceeds the maximum word length. (If it succeeds instead, it returns an `Option<MatchingWord>`).

Thinking about it more, maybe it would be better to ensure that all the components of the synonyms can have a corresponding `MatchingWord` before creating the list, instead of simply skipping the components that are too long. However, it is probably not that important.

Reply:
Well, first I'm not sure of what you expect by using `flat_map`:
1) If you want to filter the None values of a synonym and keep all the other values, I would suggest using [`filter_map`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map).
2) If you want to keep all the values until you encounter a None, I would suggest using [`map_while`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.map_while) (it seems equivalent than your current implementation, but I may be wrong).

However, you suggested to keep a synonyms only if all the words are valid, which make sens.
In this case, I think you can collect directly the value in an `Option<Vec<_>>`:
```rust
let synonym: Option<Vec<_>> = synonym
    .into_iter()
    .map(|syn| matching_word_cache.insert(syn, 0, false))
    .collect();
```
(didn't try 😬)